homeassistant/components/kraken/sensor.py

```diff
@@ -0,0 +1,229 @@
+                for sensor_type in SENSOR_TYPES:
+                    sensors.append(
+                        KrakenSensor(
+                            hass.data[DOMAIN],
+                            tracked_asset_pair,
+                            sensor_type,
+                        )
+                    )
+                async_add_entities(sensors, True)
+
+        # Remove devices for asset pairs which are no longer tracked
+        for device_id in existing_devices.values():
+            device_registry.async_remove_device(device_id)
+
+    await async_update_sensors(hass, config_entry)
+
+    hass.data[DOMAIN].unsub_listeners.append(
+        async_dispatcher_connect(
+            hass,
+            DISPATCH_CONFIG_UPDATED,
+            async_update_sensors,
+        )
+    )
+
+
+class KrakenSensor(CoordinatorEntity):
+    """Define a Kraken sensor."""
+
+    def __init__(
+        self,
+        kraken_data: KrakenData,
+        tracked_asset_pair: str,
+        sensor_type: Dict[str, bool],
+    ) -> None:
+        """Initialize."""
+        super().__init__(kraken_data.coordinator)
+        self.tracked_asset_pair_wsname = kraken_data.tradable_asset_pairs[
+            tracked_asset_pair
+        ]
+        self._source_asset = tracked_asset_pair.split("/")[0]
+        self._target_asset = tracked_asset_pair.split("/")[1]
+        self._sensor_type = sensor_type["name"]
+        self._enabled_by_default = sensor_type["enabled_by_default"]
+        self._unit_of_measurement = self._target_asset
+        self._device_name = f"{self._source_asset} {self._target_asset}"
+        self._name = "_".join(
+            [
+                tracked_asset_pair.split("/")[0],
+                tracked_asset_pair.split("/")[1],
+                sensor_type["name"],
+            ]
+        )
+        self._received_data_at_least_once = False
+        self._available = True
+        self._state = None
+
+    @property
+    def entity_registry_enabled_default(self):
+        """Return if the entity should be enabled when first added to the entity registry."""
+        return self._enabled_by_default
+
+    @property
+    def name(self):
+        """Return the name."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Set unique_id for sensor."""
+        return self._name.lower()
+
+    @property
+    def state(self):
+        """Return the state."""
+        return self._state
+
+    async def async_added_to_hass(self) -> None:
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        self._update_internal_state()
+
+    def _handle_coordinator_update(self):
+        self._update_internal_state()
+        return super()._handle_coordinator_update()
+
+    def _update_internal_state(self):
+        try:
+            self._state = self._try_get_state()
+            self._received_data_at_least_once = True  # Received data at least one time.
+        except TypeError:
```

Comment:
What can cause a `TypeError` ?

Reply:
A TypeError means the requested asset pair was not returned by the kraken api and is most likely no longer available